theano/compile/function_module.py

```diff
@@ -485,16 +486,180 @@ def __setitem__(self, item, value):
+                if key not in i_o_vars:
+                    new_storage_map[memo[key]] = storage_map[key]
+
+        input_storage = [i.value for i in ins]
+        # reinitialize new maker and create new function
+        f_cpy = maker.__class__(inputs=ins, outputs=outs,
+                                fgraph=fg_cpy,
+                                mode=maker.mode, profile=maker.profile,
+                                on_unused_input=maker.on_unused_input,
+                                function_builder=maker.function_builder,
+                                accept_inplace=maker.accept_inplace).create(
+                                input_storage, storage_map=new_storage_map)
+
+        # Share immutable and constant input's storage
+        for in_ori, in_cpy, ori, cpy in zip(maker.inputs, f_cpy.maker.inputs,
+                                            self.input_storage,
+                                            f_cpy.input_storage):
+            is_const = isinstance(in_ori.variable, theano.tensor.Constant)
+            # In instances' name default to vairables' name
+            swapped = swap is not None and in_ori.name in swapped_sv
+
+            if (is_const or not in_ori.mutable) and not swapped:
+                cpy.data = ori.data
+                in_cpy.value = in_ori.value
+
+        # Reconstruct Function.finder.
+        # Function.value and Function.data work
+        for ori, cpy in zip(maker.inputs, f_cpy.maker.inputs):
+            swapped = swap is not None and ori.name in swapped_sv
+            if not swapped:
+                f_cpy.finder[ori.variable] = f_cpy.finder.pop(cpy.variable)
+            else:
+                f_cpy.finder[swap[ori.name]] = f_cpy.finder.pop(cpy.variable)
+
+        return f_cpy
+
+    def __swapSV(self, swap, ins, fg_cpy):
+        """
+        Hidden auxiliary function, used to swap SharedVariable in maker.inputs
+        and maker.fgraph.
+        --------------------
+        Params:
+            swap -- { dict } The same as docs in copy()
+            ins -- { list } List of In instances to be modified.
+            fg_cpy -- { FounctionGraph } Copied FunctionGraph.
+        --------------------
+        Returns:
+            None
+        """
+        def checkSV(sv_ori, sv_rpl):
+            """
+            Assert two SharedVariable follow some restirctions:
+                1. same type
+                2. same shape or dim?
+            """
+            assert sv_ori.type == sv_rpl.type, (
+                "Type of given SharedVariable conflicts with origianl one",
+                "Type of given SharedVariable:", sv_rpl.type,
+                "Type of original SharedVariable:", sv_ori.type)
+
+        exist_names = [i.variable.name for i in ins]
+        swap_names = swap.keys()
+
+        # Check if given names exist
+        for name in swap_names:
+            if name not in exist_names:
+                warnings.warn("Given name: %s wasn't found" % (name))
```

Comment:
this should be an error. We should not swap by name, but by shared var instance. {shared_var: new_shared_var}